src/sentry/runner/commands/import_system_symbols.py

```diff
@@ -0,0 +1,134 @@
+"""
+sentry.runner.commands.import_system_symbols
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+:copyright: (c) 2015 by the Sentry Team, see AUTHORS for more details.
+:license: BSD, see LICENSE for more details.
+"""
+from __future__ import absolute_import
+
+import uuid
+import json
+import click
+import zipfile
+import threading
+import Queue
+from django.db import connection, IntegrityError
+from sentry.runner.decorators import configuration
+
+
+SHUTDOWN = object()
+
+
+def load_bundle(q, uuid, data, sdk_info):
+    from sentry.models import DSymBundle, DSymObject, DSymSDK
+
+    sdk = DSymSDK.objects.get_or_create(
+        dsym_type=sdk_info['dsym_type'],
+        sdk_name=sdk_info['sdk_name'],
+        version_major=sdk_info['version_major'],
+        version_minor=sdk_info['version_minor'],
+        version_patchlevel=sdk_info['version_patchlevel'],
+        version_build=sdk_info['version_build'],
+    )[0]
+
+    obj = DSymObject.objects.get_or_create(
+        cpu_name=data['arch'],
+        object_path='/' + data['image'].strip('/'),
+        uuid=str(uuid),
+        vmaddr=data['vmaddr'],
+        vmsize=data['vmsize'],
+    )[0]
+
+    DSymBundle.objects.get_or_create(
+        sdk=sdk,
+        object=obj
+    )[0]
+
+    step = 4000
+    symbols = data['symbols']
+    for idx in xrange(0, len(symbols) + step, step):
+        end_idx = min(idx + step, len(symbols))
+        yield [{
+            'object_id': obj.id,
+            'address': symbols[x][0],
+            'symbol': symbols[x][1],
+        } for x in xrange(idx, end_idx)]
+
+
+def process_archive(members, zip, sdk_info, threads):
+    q = Queue.Queue(threads)
+
+    def process_items():
+        cur = connection.cursor()
+        cur.execute('begin')
+        cur.execute('''
+            prepare add_sym(bigint, bigint, text) as
+                insert into sentry_dsymsymbol (object_id, address, symbol)
+                select $1, $2, $3
+                where not exists (select 1 from sentry_dsymsymbol
+                    where object_id = $1 and address = $2);
+        ''')
+        while 1:
+            items = q.get()
+            if items is SHUTDOWN:
+                break
+            try:
+                cur.executemany('''
+                    execute add_sym(%(object_id)s, %(address)s, %(symbol)s);
+                ''', items)
+            except IntegrityError:
```

Comment:
This is going to abort the entire transaction, no? and every subsequent EXECUTE is going to fail.